### PR TITLE
Update example for bounding volume hierarchy 

### PR DESCRIPTION
**Pull request by dalg24**
* Update example bvh driver and use it for measuring performance
* Make it run as part of the automated testing

### CODE REVIEW

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 1]
```diff
@@ -95,110 +47,99 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
-        bounding_boxes_host[i] = {{{x, y, z}}, {{x, y, z}}};
+        auto n = std::max( n_values, n_queries );
+        Kokkos::resize( random_points, n );
+
+        auto random_points_host = Kokkos::create_mirror_view( random_points );
+
+        // edge length chosen such that object density will remain constant as
+        // problem size is changed
+        auto const a = std::cbrt( n_values );
+        std::uniform_real_distribution<double> distribution( -a, +a );
```
Comment (sslattery) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: So here you are sampling a 3D box that spans '(-a,+a)' in each direction to get the points. You use the same points for queries and construction hence the `std:max` operation to compute `n`.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 2]
```diff
@@ -95,110 +47,99 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
+                              bounding_boxes( i ) = {
+                                  {{x - 1., y - 1., z - 1.}},
+                                  {{x + 1., y + 1., z + 1.}}};
+                          } );
+    Kokkos::fence();
+
+    std::ostream &os = std::cout;
+
+    auto start = std::chrono::high_resolution_clock::now();
     DataTransferKit::BVH<DeviceType> bvh( bounding_boxes );
```
Comment (sslattery) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Does Kokkos automatically launch a little kernel to "warm up" the device? I know there is always some type of overhead with the first kernel launch in a program. This likely does not even come close to accounting for the scaling of the construction using CUDA but would be interesting to know. If I remember I think they may call a small no-op kernel upon calling initialize.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 3]
```diff
@@ -95,110 +47,99 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
-
-        Kokkos::View<details::Within *, DeviceType> within_queries(
-            "within_queries", n_points );
+    {
+        Kokkos::View<DataTransferKit::Details::Within *, DeviceType> queries(
+            "queries", n_queries );
+        // radius chosen such that there will be approximately 10 results per
+        // query
+        double const pi = 3.14159265359;
+        double const r = std::cbrt( 10. * 3. / ( 4. * pi ) );
```
Comment (sslattery) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Given that the input point set is random I think this is about as good as you can do to get an equivalent number of numbers from both kNN and radius search.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 4]
```diff
@@ -95,110 +47,99 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
+        Kokkos::View<int *, DeviceType> indices( "indices" );
+        start = std::chrono::high_resolution_clock::now();
+        bvh.query( queries, indices, offset );
+        end = std::chrono::high_resolution_clock::now();
+        elapsed_seconds = end - start;
+        os << "radius " << elapsed_seconds.count() << "\n";
+        // check that they are indeed 10 results per query (within 15% is fine)
+        DTK_ENSURE(
+            boost::math::relative_difference(
+                static_cast<double>( DataTransferKit::lastElement( offset ) ) /
```
Comment (sslattery) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Has this ever failed in a DBC build?

Comment (dalg24) [reply packages/Search/examples/bvh_driver/bvh_driver.cpp]: Yup it failed with 1 billion values and 1 billion queries.  The truth is we do intersection of a sphere of radius 10 with boxes 2x2x2.  So the actual radius for interaction is `10+sqrt(3)` which sums up to `11.73`...  oops

File: packages/Search/examples/bvh_driver/bvh_driver.cpp (hunk 1 above)
Comment (dalg24) [reply packages/Search/examples/bvh_driver/bvh_driver.cpp]: Correct.

**Comment by dalg24** [comment]
I still have not been able to reproduce [these failures](https://cloud.cees.ornl.gov/jenkins-ci/job/DataTransferKit-Clang50/305/) locally...

File: packages/Search/examples/bvh_driver/bvh_driver.cpp (hunk 2 above)
Comment (dalg24) [reply packages/Search/examples/bvh_driver/bvh_driver.cpp]: See [`Kokkos::Impl::CudaInternal::initialize()`](https://github.com/kokkos/kokkos/blob/master/core/src/Cuda/Kokkos_Cuda_Impl.cpp#L370-L564).

**Comment by dalg24** [comment]
> It would be nice to optionally change the number of neighbors to query for or a radius as an input parameter so we can vary that as well. In fact, increasing the number of neighbors in a query is another important algorithmic scaling plot to make as our users will have varying query sizes.

Now you can pass `--neighbors=<K>` to change the default value.

File: packages/Search/examples/bvh_driver/CMakeLists.txt [hunk 5]
```diff
@@ -5,5 +5,33 @@
...
+    )
+ENDIF()
+IF(Kokkos_ENABLE_OpenMP)
+  TRIBITS_ADD_TEST(
+    bvh
+    POSTFIX_AND_ARGS_0 openmp --node=openmp
+    COMM serial mpi
+    NUM_MPI_PROCS 1
+    NUM_TOTAL_CORES_USED 4
+    ENVIRONMENT OMP_NUM_THREADS=4
```
Comment (dalg24) [inline packages/Search/examples/bvh_driver/CMakeLists.txt]: c.f. #375

**Comment by dalg24** [comment]
Ready for final review/merge

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 6]
```diff
@@ -8,80 +8,31 @@
...
+    int n_neighbors = 10;
 
-    clp.setOption( "nx", &nx, "source mesh points in x-direction." );
-    clp.setOption( "ny", &ny, "source mesh points in y-direction." );
-    clp.setOption( "nz", &nz, "source mesh points in z-direction." );
-    clp.setOption( "N", &n_points,
-                   "number of target mesh points (distributed randomly)." );
-    clp.setOption( "mode", &mode, "mode: (knn | radius)" );
+    clp.setOption( "values", &n_values, "number of indexable values (source)" );
+    clp.setOption( "queries", &n_queries, "number of queries (target)" );
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: I think the language is a bit confusing. I tend to think about the problem in source point and target points. But you think in terms of values and queries. That's foreign to me, but probably fine.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 7]
```diff
@@ -95,110 +46,97 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
+    Kokkos::View<DataTransferKit::Point *, DeviceType> random_points(
+        "random_points" );
     {
-        auto const &point = cloud[i];
-        double x = std::get<0>( point );
-        double y = std::get<1>( point );
-        double z = std::get<2>( point );
-        bounding_boxes_host[i] = {{{x, y, z}}, {{x, y, z}}};
+        auto n = std::max( n_values, n_queries );
+        Kokkos::resize( random_points, n );
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: It looks a bit weird. You have a view but size it as max of number of source points and number of target points. I would have expected to create those independently.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 8]
```diff
@@ -95,110 +46,97 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
+
+        auto random_points_host = Kokkos::create_mirror_view( random_points );
+
+        // edge length chosen such that object density will remain constant as
+        // problem size is changed
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Maybe rephrase the comment? What edge? What object density? Be more verbose.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 9]
```diff
@@ -95,110 +46,97 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
+        auto const a = std::cbrt( n_values );
+        std::uniform_real_distribution<double> distribution( -a, +a );
+        std::default_random_engine generator;
+        auto random = [&distribution, &generator]() {
+            return distribution( generator );
+        };
+        for ( int i = 0; i < n; ++i )
+            random_points_host( i ) = {{random(), random(), random()}};
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: This should be reproducible, right? I.e., the seed is going to be the same.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 10]
```diff
@@ -95,110 +46,97 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
 
+    Kokkos::View<DataTransferKit::Box *, DeviceType> bounding_boxes(
+        "bounding_boxes", n_values );
+    Kokkos::parallel_for( Kokkos::RangePolicy<ExecutionSpace>( 0, n_values ),
+                          KOKKOS_LAMBDA( int i ) {
+                              double const x = random_points( i )[0];
+                              double const y = random_points( i )[1];
+                              double const z = random_points( i )[2];
+                              bounding_boxes( i ) = {
+                                  {{x - 1., y - 1., z - 1.}},
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Make 1 a tweakable parameter.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 11]
```diff
@@ -95,110 +46,97 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
-        double z = std::get<2>( point );
-        bounding_boxes_host[i] = {{{x, y, z}}, {{x, y, z}}};
+        auto n = std::max( n_values, n_queries );
+        Kokkos::resize( random_points, n );
+
+        auto random_points_host = Kokkos::create_mirror_view( random_points );
+
+        // edge length chosen such that object density will remain constant as
+        // problem size is changed
+        auto const a = std::cbrt( n_values );
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Should you also allow for 2D problems in this test? This would require changing `cbrt` to `sqrt` and some other things in it.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 12]
```diff
@@ -95,110 +46,97 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
-            Kokkos::RangePolicy<ExecutionSpace>( 0, n_points ),
+            Kokkos::RangePolicy<ExecutionSpace>( 0, n_queries ),
             KOKKOS_LAMBDA( int i ) {
-                nearest_queries( i ) = details::nearest<DataTransferKit::Point>(
-                    {{point_coords( i, 0 ), point_coords( i, 1 ),
-                      point_coords( i, 2 )}},
-                    k( i ) );
+                queries( i ) =
+                    DataTransferKit::Details::nearest<DataTransferKit::Point>(
+                        random_points( i ), n_neighbors );
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: So your target points coincide with your source points? Was that done for simplicity?

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 13]
```diff
@@ -95,110 +46,97 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
-        Kokkos::View<details::Within *, DeviceType> within_queries(
-            "within_queries", n_points );
+    {
+        Kokkos::View<DataTransferKit::Details::Within *, DeviceType> queries(
+            "queries", n_queries );
+        // radius chosen in order to control the number of results per query
+        // NOTE: minus "1+sqrt(3)/2 \approx 1.37" matches the size of the boxes
+        // inserted into the tree (mid-point between half-edge and
+        // half-diagonal)
+        double const pi = 3.14159265359;
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Use `M_PI` from `cmath`.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 14]
```diff
@@ -95,110 +46,97 @@ int main_( Teuchos::CommandLineProcessor &clp, int argc, char *argv[] )
...
+                    DataTransferKit::Details::within( random_points( i ), r );
             } );
         Kokkos::fence();
 
-        Kokkos::View<int *, DeviceType> offset_within( "offset_within" );
-        Kokkos::View<int *, DeviceType> indices_within( "indices_within" );
-        bvh.query( within_queries, indices_within, offset_within );
+        Kokkos::View<int *, DeviceType> offset( "offset" );
+        Kokkos::View<int *, DeviceType> indices( "indices" );
+        start = std::chrono::high_resolution_clock::now();
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Can you add blank line before this one? I think it would be more readable to have a separate code block like
```c++
Kokkos::View<int *, DeviceType> offset( "offset" );
Kokkos::View<int *, DeviceType> indices( "indices" );

start = std::chrono::high_resolution_clock::now();
bvh.query( queries, indices, offset );
end = std::chrono::high_resolution_clock::now();

elapsed_seconds = end - start;
os << "radius " << elapsed_seconds.count() << "\n";
```

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 15]
```diff
@@ -225,27 +162,27 @@ int main( int argc, char *argv[] )
...
-        if ( rv )
+        if ( !success )
         {
             // do nothing, just skip other if clauses
         }
         else if ( node == "" )
         {
             typedef KokkosClassic::DefaultNode::DefaultNodeType Node;
-            rv = main_<Node>( clp, argc, argv );
+            main_<Node>( clp, argc, argv );
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: If you are not using the return value from `main_`, might as well change it to `void`.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 16]
```diff
@@ -8,80 +8,31 @@
...
+    int n_values = 50000;
+    int n_queries = 20000;
+    int n_neighbors = 10;
 
-    clp.setOption( "nx", &nx, "source mesh points in x-direction." );
-    clp.setOption( "ny", &ny, "source mesh points in y-direction." );
-    clp.setOption( "nz", &nz, "source mesh points in z-direction." );
-    clp.setOption( "N", &n_points,
-                   "number of target mesh points (distributed randomly)." );
-    clp.setOption( "mode", &mode, "mode: (knn | radius)" );
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Please keep the `mode` option in. If I want to do some scaling studies just for `knn`, there is no reason for the simulation to take twice as long doing unnecessary `radius` searches.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp (hunk 8 above)
Comment (dalg24) [reply packages/Search/examples/bvh_driver/bvh_driver.cpp]: I updated the comment.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp (hunk 16 above)
Comment (dalg24) [reply packages/Search/examples/bvh_driver/bvh_driver.cpp]: I added boolean options to decide whether to run kNN or radius search.  Both are running by default.
